src/emit/substrate/target.rs

```diff
@@ -1580,14 +1580,73 @@ impl<'a> TargetRuntime<'a> for SubstrateTarget {
 
     fn builtin_function(
         &self,
-        _binary: &Binary<'a>,
-        _function: FunctionValue<'a>,
-        _builtin_func: &Function,
-        _args: &[BasicMetadataValueEnum<'a>],
+        binary: &Binary<'a>,
+        function: FunctionValue<'a>,
+        builtin_func: &Function,
+        args: &[BasicMetadataValueEnum<'a>],
         _first_arg_type: BasicTypeEnum,
         _ns: &Namespace,
     ) -> BasicValueEnum<'a> {
-        unimplemented!()
+        emit_context!(binary);
+
+        assert_eq!(builtin_func.name, "chain_extension", "unimplemented");
+
+        let input_ptr = binary.vector_bytes(args[1].into_pointer_value().into());
+        let input_len = binary.vector_len(args[1].into_pointer_value().into());
+        let (output_ptr, output_len_ptr) = scratch_buf!();
+        let len = 16384;
+        binary.builder.build_store(output_len_ptr, i32_const!(len));
+        call!(
+            "__memset8",
+            &[
+                output_ptr.into(),
+                binary.context.i64_type().const_zero().into(),
+                i32_const!(len / 8).into()
+            ]
+        );
+        let ret = call!(
+            "call_chain_extension",
+            &[
+                args[0].into_int_value().into(),
+                input_ptr.into(),
+                input_len.into(),
+                output_ptr.into(),
+                output_len_ptr.into()
+            ]
+        )
+        .try_as_basic_value()
+        .left()
+        .unwrap()
+        .into_int_value();
+
+        log_return_code(binary, "call_chain_extension", ret);
+
+        let ok = binary.context.append_basic_block(function, "ok_block");
+        let done = binary.context.append_basic_block(function, "done_block");
+        let success = binary
+            .builder
+            .build_int_compare(IntPredicate::EQ, ret, i32_zero!(), "ret");
+        binary.builder.build_conditional_branch(success, ok, done);
+
+        binary.builder.position_at_end(ok);
+        let buf_len = binary
+            .builder
+            .build_load(binary.context.i32_type(), output_len_ptr, "buf_len")
+            .into_int_value();
+        let buf = call!(
+            "vector_new",
+            &[buf_len.into(), i32_const!(1).into(), output_ptr.into(),]
+        )
+        .try_as_basic_value()
+        .left()
+        .unwrap();
+        binary
+            .builder
+            .build_store(args[2].into_pointer_value(), buf.into_pointer_value());
+        binary.builder.build_unconditional_branch(done);
+
+        binary.builder.position_at_end(done);
+        ret.into()
```

Comment:
This looks wrong. Why is the syscall return value returned and not the vector?

Reply:
Are you sure? [An int value is expected](https://github.com/hyperledger/solang/blob/main/src/emit/instructions.rs#LL532C23-L532C23), a pointer won't work. I guessed that this int value is supposed to be the return code from the builtin.

The return value goes [into the last argument to the builtin](https://github.com/hyperledger/solang/blob/main/src/emit/instructions.rs#L525). I'm quite confident that I read this code correctly and implemented it accordingly. The unit test passes, and out of curiosity I tried it out on a real substrate node with the `fetch-rand` chain extension example. It works as expected.